### ISSUE TITLE
 NAN of batchnorm in AMP

### ISSUE THREAD

**Issue by SakuraRiven** [pytorch/pytorch#96798]
Hi, we have a model where some layers are
```
...
self.shared_1 = nn.Conv2d(512, 64, kernel_size=3, padding=1, bias=False)
self.shared_2 = nn.Batchnorm2d(64, eps=1e-3)
self.shared_3 = nn.ReLU(True)
...
```
The forward is
```
x1 = self.shared_1(x)
x2 = self.shared_2(x1)
x3 = self.shared_3(x2)
```
When we using AMP, we found that the __running_mean__ and __runnning_var__ of ```self.shared_2``` are normal before receiving x1. After producing x2, however, there are some NANs in its  __running_mean__ and __runnning_var__. Maybe these values calculated from x1 exceed the range of half precision?

Are there any elegant approach to avoid NAN? Thanks

For instance, how could we keep above conv layer as FP32?

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @mcarilli @ptrblck @leslie-fang-intel @jgong5

**Comment by mingfeima** [comment]
@SakuraRiven in the AMP mode, `BatchNorm` will keep running_mean and running_var in float32. it is possible that the overflow happens when updating parameters.

Could you please provide a minimal example to reproduce your issue? 

you can explicitly convert the output of shared_1 to float in your `forward`, then batch_norm will use fp32 for computation (input/output, parameter all in fp32), but the input to `ReLU` will also be fp32.

**Comment by SakuraRiven** [comment]
@mingfeima The entire model has large architecture, where the ```backbone```, ```neck``` have been checked normal.

The above-mentioned layers are in the ```head``` part, which starts to appear nan.

Maybe we could find more detailed bug layer by progressively replacing FP16 layers as FP32? If so, we know that the autocast(enable=False) could be used in forward() to set specific layers as FP32. Are there some approaches to keep specific layers as FP32 in __init__ function?

**Comment by vadimkantorov** [comment]
One option may be manual updates of batchnorm running stats

**Comment by ptrblck** [comment]
Cross-post from [here](https://discuss.pytorch.org/t/nan-of-batchnorm-in-amp/174877/2?u=ptrblck) with additional discussion.

**Comment by SakuraRiven** [comment]
So confused ...
1. We found that amp is successfully when only __one gpu__ is used
2. The first invalid value appears in the __x1__, which is the output of ```conv``` and contains __inf__

**Comment by ezhang887** [comment]
> Solved.

Hi @SakuraRiven how did you solve the issue? I may be seeing a similar issue in one of my workloads.

**Comment by SakuraRiven** [comment]
@ezhang887  It is actually a special bug in our framework which employs complicated wrapper on optimizer and the AMP setting does not take effect.

As for the Nan problem, maybe this could help you https://discuss.pytorch.org/t/nan-loss-with-torch-cuda-amp-and-crossentropyloss/108554/19
